## Instructions {-}

1. You may talk to a friend, discuss the questions and potential directions for solving them. However, you need to write your own solutions and code separately, and not as a group activity. 

2. Do not write your name on the assignment.

3. Write your code in the *Code* cells and your answer in the *Markdown* cells of the Jupyter notebook. Ensure that the solution is written neatly enough to understand and grade.

4. Use [Quarto](https://quarto.org/docs/output-formats/html-basics.html) to print the *.ipynb* file as HTML. You will need to open the command prompt, navigate to the directory containing the file, and use the command: `quarto render filename.ipynb --to html`. Submit the HTML file.

5. The assignment is worth 100 points, and is due on **22nd October 2023 at 11:59 pm**. There is an optional bonus question worth 10 points. You can score a maximum of 110 (out of 100) points.

6. You are **not allowed to use a `for` loop** or any other kind of loop in this assignment.

## GDP per capita & social indicators
Read the file *social_indicator.txt* with python. Set the first column as the index when reading the file. How many observations and variables are there in the data?

*(4 points)*

## C.1 ##

In [1]:
#C.1.1
import pandas as pd

file_path = '/Users/jinwens/Downloads/social_indicator.txt'
df = pd.read_csv(file_path, delimiter="\t", index_col=0)

df.info()

df.head()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 155 entries, 1 to 36
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   economicActivityFemale  155 non-null    float64
 1   country                 155 non-null    object 
 2   region                  155 non-null    object 
 3   totalfertilityrate      155 non-null    float64
 4   lifeMale                155 non-null    float64
 5   infantMortality         155 non-null    int64  
 6   gdpPerCapita            155 non-null    int64  
 7   economicActivityMale    155 non-null    float64
 8   illiteracyMale          155 non-null    float64
 9   illiteracyFemale        155 non-null    float64
 10  lifeFemale              155 non-null    float64
 11  geographic_location     155 non-null    object 
 12  contraception           144 non-null    float64
dtypes: float64(8), int64(2), object(3)
memory usage: 17.0+ KB


economicActivityFemale      country region  totalfertilityrate  \
Index                                                                   
1                         1.9        Yemen  other                 7.6   
9                        15.8         Oman  other                 7.2   
124                      58.4     Ethiopia  other                 7.0   
3                         7.2  Afghanistan  other                 6.9   
11                       20.2     Maldives  other                 6.8   

       lifeMale  infantMortality  gdpPerCapita  economicActivityMale  \
Index                                                                  
1          57.4               80          1924                  80.6   
9          68.9               25         30404                  84.1   
124        48.4              107          2973                  84.7   
3          45.0              154          2474                  87.5   
11         65.7               49         23343                  77.3   

       illiteracyMale  illiteracyFemale  lifeFemale geographic_location  \
Index                                                                     
1              32.406            69.552        58.4        Western Asia   
9               3.010             7.310        73.3        Western Asia   
124            54.500            74.700        51.6      Eastern Africa   
3              52.800            85.000        46.0       Southern Asia   
11              6.700             7.000        63.3       Southern Asia   

       contraception  
Index                 
1                0.0  
9               53.0  
124             79.0  
3                NaN  
11              53.0

The dataset contains
155 observations and 13 variables, including social indicators and geographical information.



### 
Which variables have the strongest and weakest correlations with GDP per capita? Note that `lifeFemale` and `lifeMale` are the female and male life expectancies respectively.

Note that only when the magnitude of the correlation is considered when judging a correlation as strong or weak.

*(4 points)*

In [2]:
#C1.2
correlation_with_gdp = df.corr()['gdpPerCapita'].drop('gdpPerCapita').abs().sort_values(ascending=False)
correlation_with_gdp


/var/folders/vh/bnk8gjpn5998vq44hzdttvwc0000gn/T/ipykernel_19537/2491558818.py:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  correlation_with_gdp = df.corr()['gdpPerCapita'].drop('gdpPerCapita').abs().sort_values(ascending=False)


lifeFemale                0.604029
lifeMale                  0.592267
infantMortality           0.584060
totalfertilityrate        0.548172
illiteracyMale            0.471689
illiteracyFemale          0.457012
economicActivityMale      0.167231
economicActivityFemale    0.052964
contraception             0.004286
Name: gdpPerCapita, dtype: float64

The variable that has the strongest correlation with GDP per capita is lifeFemale(female life expectancies), with a correlation coefficient of approximately **0.604029**.
The variable with the weakest correlation with GDP per capita is contraception, with a correlation coefficient of approximately **0.004286**.



### 
Does the male economic activity *(in the column `economicActivityMale`)* have a positive or negative correlation with GDP per capita? Did you expect the positive/negative correlation? If not, why do you think you are observing that correlation?

*(4 points)*

In [3]:
#C1.3

The male economic activity has a positive correlation with the GDP per capita with a coefficient being 0.167231. This indicates a weak positive correlation: when the male economic activity increases, the GDP per capita also tends to increase slightly. I did expect this positive correlation. 

### 

What is the rank of the US amongst all countries in terms of GDP per capita? Which countries lie immediately above, and immediately below the US in the ranking in terms of GDP per capita? The country having the highest GDP per capita ranks 1.

Note that:

1. The US is mentioned as *United States* in the data. 

2. The country with the highest GDP per capita will have rank 1, the country with the second highest GDP per capita will have rank 2, and so on.

**Hint:** [rank()](https://pandas.pydata.org/docs/reference/api/pandas.Series.rank.html)

*(4 points)*

In [5]:
# C1.3
df['gdp_rank'] = df['gdpPerCapita'].rank(method='min', ascending=False)

us_rank = df[df['country'] == 'United States']['gdp_rank'].iloc[0]

# Finding the countries that are immediately above and below the United States
above_us = df[df['gdp_rank'] == us_rank - 1]['country'].iloc[0]
below_us = df[df['gdp_rank'] == us_rank + 1]['country'].iloc[0]

(us_rank, above_us, below_us)


(8.0, 'Norway', 'Brunei')

The United States ranks 8th among the countries in the dataset. Norway is immediately above the United States, ranking 7th, and Brunei is immediately below, ranking 9th.

### 

Which country or countries rank among the top 20 in terms of each of these social indicators - `economicActivityFemale`, `economicActivityMale`, `gdpPerCapita`, `lifeFemale`, `lifeMale`? For each of these social indicators, the country having the largest value ranks 1 for that indicator.

*(6 points)*

**Hint:** 

1. Use [rank()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rank.html). Note that this method is different from the method given in the hint of the previous question. This method is of the DataFrame class, while the one in the previous question is of the Series class. *This part of the hint is just for your understanding. You don't need to write any code in this part.*

2. Using `rank()`, get the DataFrame consisting of the ranks of countries on each of the relevant social indicators *(one line of code)*.

3. In the DataFrame obtained in (2), filter the rows for which the maximum rank is less than or equal to 20 *(one line of code)*.

In [6]:
#C1.4
ranked_df = df[['economicActivityFemale', 'economicActivityMale', 'gdpPerCapita', 'lifeFemale', 'lifeMale']].rank(ascending=False)

# Filtering the rows for which the maximum rank is less than or equal to 20
top_20_countries = df[ranked_df.max(axis=1) <= 20]
top_20_countries


economicActivityFemale  country region  totalfertilityrate  lifeMale  \
Index                                                                         
149                      76.8  Iceland   west                2.19      77.4   

       infantMortality  gdpPerCapita  economicActivityMale  illiteracyMale  \
Index                                                                        
149                  5         58151                  86.2             1.0   

       illiteracyFemale  lifeFemale geographic_location  contraception  \
Index                                                                    
149                 1.0        81.2     Northern Europe           77.0   

       gdp_rank  
Index            
149        13.0

Only Iceland ranks among the top 20 countries in terms of each of the social indicators: economicActivityFemale, economicActivityMale, gdpPerCapita, lifeFemale, and lifeMale.

### 

On which social indicator among `economicActivityFemale`, `economicActivityMale`, `gdpPerCapita`, `lifeFemale`, `lifeMale`, `illiteracyFemale`, `illiteracyMale`, `infantMortality`, and `totalfertilityrate` does the US have its worst ranking, and what is the rank? Note that for `illiteracyFemale`, `illiteracyMale`, and `infantMortality`, the country having the lowest value will rank 1, in contrast to the other social indicators.

*(8 points)*

In [11]:
ranking_df = df[['economicActivityFemale', 'economicActivityMale', 'gdpPerCapita', 'lifeFemale', 'lifeMale', 'totalfertilityrate']].rank(ascending=False)

ranking_df[['illiteracyFemale', 'illiteracyMale', 'infantMortality']] = df[['illiteracyFemale', 'illiteracyMale', 'infantMortality']].rank(ascending=True)

us_ranks = ranking_df.loc[df['country'] == 'United States']  

worst_rank = us_ranks.idxmax(axis=1).values[0]
worst_rank_value = us_ranks.max(axis=1).values[0]

worst_rank, worst_rank_value

('totalfertilityrate', 109.0)

The United States has the worst-ranking social indicator being totalfertilityrate, where it ranks 109th among the countries in the dataset.

### 

Find all the countries that have a lower GDP per capita than the US, despite having lower illiteracy rates (for both genders), higher economic activity (for both genders), higher life expectancy (for both genders), and lower infant mortality rate than the US?

*(6 points)*

In [52]:
# C1.6
us_values = df[df['country'] == 'United States'][['gdpPerCapita', 'illiteracyMale', 'illiteracyFemale',
                                                 'economicActivityMale', 'economicActivityFemale', 
                                                 'lifeMale', 'lifeFemale', 'infantMortality']].iloc[0]

better_countries = df[
    (df['gdpPerCapita'] < us_values['gdpPerCapita']) &
    (df['illiteracyMale'] < us_values['illiteracyMale']) &
    (df['illiteracyFemale'] < us_values['illiteracyFemale']) &
    (df['economicActivityMale'] > us_values['economicActivityMale']) &
    (df['economicActivityFemale'] > us_values['economicActivityFemale']) &
    (df['lifeMale'] > us_values['lifeMale']) &
    (df['lifeFemale'] > us_values['lifeFemale']) &
    (df['infantMortality'] < us_values['infantMortality'])
]


better_countries['country']


Index
149        Iceland
147         Sweden
129    Netherlands
Name: country, dtype: object

The countries have a lower GDP per capita than the United States, yet perform better in terms of lower illiteracy rates (for both genders), higher economic activity (for both genders), higher life expectancy (for both genders), and lower infant mortality rate are:

Iceland
Sweden
Netherlands

## GDP per capita vs social indicators
We'll use the same data as in in the previous question. For the questions below, assume that all numeric columns, **except GDP per capita**, are social indicators.

###  
Use the column `geographic_location` to create a new column called `continent`. Merge the values of the `geographic_location` column appropriately to obtain 6 distinct values for the `continent` column – *Asia, Africa, North America, South America, Europe* and *Oceania*. Drop the column `geographic_location`. Print the first 5 observations of the updated DataFrame.

*(8 points)*

**Hint:**

1. Use `value_counts()` to see the values of `geographic_location`. The code `if 'Asia' in 'something'` will return `True` if '*something*' contains the string '*Asia*', for example, if '*something*' is '*North Asia*', the code with return True. *This part of the hint is just for your understanding. You don't need to write any code in this part.*

2. Apply a lambda function on the Series `geographic_location` to replace a string that contains '*Asia*' with '*Asia*', replace a string that contains '*Europe*' with '*Europe*', and replace a string that contains '*Africa*' with '*Africa*'. *This will be a single line of code.*

3. Rename the column `georgaphic_location` to `continent`.

In [54]:
# C2.1
df['continent'] = df['geographic_location'].apply(
    lambda x: 'Asia' if 'Asia' in x else (
        'Africa' if 'Africa' in x else (
            'North America' if 'North America' in x else (
                'South America' if 'South America' in x else (
                    'Europe' if 'Europe' in x else 'Oceania')))))

df = df.drop(columns=['geographic_location'])

df.head()


economicActivityFemale      country region  totalfertilityrate  \
Index                                                                   
1                         1.9        Yemen  other                 7.6   
9                        15.8         Oman  other                 7.2   
124                      58.4     Ethiopia  other                 7.0   
3                         7.2  Afghanistan  other                 6.9   
11                       20.2     Maldives  other                 6.8   

       lifeMale  infantMortality  gdpPerCapita  economicActivityMale  \
Index                                                                  
1          57.4               80          1924                  80.6   
9          68.9               25         30404                  84.1   
124        48.4              107          2973                  84.7   
3          45.0              154          2474                  87.5   
11         65.7               49         23343                  77.3   

       illiteracyMale  illiteracyFemale  lifeFemale  contraception  gdp_rank  \
Index                                                                          
1              32.406            69.552        58.4            0.0     149.0   
9               3.010             7.310        73.3           53.0      48.0   
124            54.500            74.700        51.6           79.0     138.0   
3              52.800            85.000        46.0            NaN     144.0   
11              6.700             7.000        63.3           53.0      59.0   

      continent  
Index            
1          Asia  
9          Asia  
124      Africa  
3          Asia  
11         Asia

### 
Sort the column labels lexicographically. Drop the columns `region` and `contraception`. Print the first 5 observations of the updated DataFrame.

**Hint:** [sort_index()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_index.html)

*(4 points)*

In [55]:
#C 2.2
df = df.sort_index(axis=1)

df = df.drop(columns=['region', 'contraception'])

df.head()



continent      country  economicActivityFemale  economicActivityMale  \
Index                                                                        
1          Asia        Yemen                     1.9                  80.6   
9          Asia         Oman                    15.8                  84.1   
124      Africa     Ethiopia                    58.4                  84.7   
3          Asia  Afghanistan                     7.2                  87.5   
11         Asia     Maldives                    20.2                  77.3   

       gdpPerCapita  gdp_rank  illiteracyFemale  illiteracyMale  \
Index                                                             
1              1924     149.0            69.552          32.406   
9             30404      48.0             7.310           3.010   
124            2973     138.0            74.700          54.500   
3              2474     144.0            85.000          52.800   
11            23343      59.0             7.000           6.700   

       infantMortality  lifeFemale  lifeMale  totalfertilityrate  
Index                                                             
1                   80        58.4      57.4                 7.6  
9                   25        73.3      68.9                 7.2  
124                107        51.6      48.4                 7.0  
3                  154        46.0      45.0                 6.9  
11                  49        63.3      65.7                 6.8

### 

Find the percentage of the total countries in each continent.

**Hint:** One line of code with `value_counts()` and `shape`

*(4 points)*

In [29]:
# C2.3
continent_percentage = (df['continent'].value_counts() / df.shape[0]) * 100
continent_percentage


Asia             26.451613
Europe           22.580645
Africa           20.000000
North America    13.548387
Oceania           9.677419
South America     7.741935
Name: continent, dtype: float64

### 

Which country has the highest GDP per capita? Let us call it country $G$.

*(4 points)*

In [56]:
# C2.4
country_highest_gdp = df.loc[df['gdpPerCapita'].idxmax()]['country']
country_highest_gdp

'Luxembourg'

The country G with the highest GDP per capita is Luxembourg.

### 
We need to find the African country that is the closest to country $G$ with regard to social indicators. Perform the following steps:

#### 	
Standardize each of the social indicators to a standard normal distribution so that all of them are on the same scale (remember to exclude *GDP per capita* from social indicators).

**Hint:**

1. For scaling a random variable to standard normal, subtract the mean from each value of the variable, and divide by its standard deviation.

2. Use the apply method with a lambda function to scale all the social indicators to standard normal.

3. The above (1) and (2) together is a single line of code.

*(6 points)*

In [61]:
social_indicators = ['economicActivityFemale', 'economicActivityMale', 'illiteracyFemale', 
                    'illiteracyMale', 'infantMortality', 'lifeFemale', 'lifeMale', 'totalfertilityrate']

df[social_indicators] = df[social_indicators].apply(lambda x: (x - x.mean()) / x.std())

df.head()



continent      country  economicActivityFemale  economicActivityMale  \
Index                                                                        
1          Asia        Yemen               -2.608875              0.521297   
9          Asia         Oman               -1.794841              0.975454   
124      Africa     Ethiopia                0.699967              1.053310   
3          Asia  Afghanistan               -2.298488              1.416635   
11         Asia     Maldives               -1.537162              0.093092   

       gdpPerCapita  gdp_rank  illiteracyFemale  illiteracyMale  \
Index                                                             
1              1924     149.0          1.910795        1.110789   
9             30404      48.0         -0.560149       -0.647801   
124            2973     138.0          2.115165        2.432543   
3              2474     144.0          2.524065        2.330842   
11            23343      59.0         -0.572456       -0.427050   

       infantMortality  lifeFemale  lifeMale  totalfertilityrate  
Index                                                             
1             1.247233   -1.198493 -0.895118            2.719042  
9            -0.333597    0.278282  0.387065            2.471951  
124           2.023276   -1.872458 -1.898566            2.348405  
3             3.374166   -2.427488 -2.277646            2.286632  
11            0.356220   -0.712842  0.030283            2.224859

#### 
Compute the [Manhattan distance](https://en.wikipedia.org/wiki/Taxicab_geometry) between country $G$ and each of the African countries, based on the scaled social indicators.

**Hint:** 

1. Broadcast a Series to a DataFrame

2. The Manhattan distance between two points $(x_1, x_2, ..., x_p)$ and $(y_1, y_2, ..., y_p)$ is $|x_1 - y_1| + |x_2 - y_2| + ... + |x_p-y_p|$, where $|.|$ stands for absolute value (for example, $|-2| = 2;  |3| = 3$).

*(8 points)*

In [69]:
# C2.5.1
social_indicators = ['economicActivityFemale', 'economicActivityMale', 'illiteracyFemale', 
                    'illiteracyMale', 'infantMortality', 'lifeFemale', 'lifeMale', 'totalfertilityrate']

df[social_indicators] = df[social_indicators].apply(lambda x: (x - x.mean()) / x.std())

df.head()


continent      country  economicActivityFemale  economicActivityMale  \
Index                                                                        
1          Asia        Yemen               -2.608875              0.521297   
9          Asia         Oman               -1.794841              0.975454   
124      Africa     Ethiopia                0.699967              1.053310   
3          Asia  Afghanistan               -2.298488              1.416635   
11         Asia     Maldives               -1.537162              0.093092   

       gdpPerCapita  gdp_rank  illiteracyFemale  illiteracyMale  \
Index                                                             
1              1924     149.0          1.910795        1.110789   
9             30404      48.0         -0.560149       -0.647801   
124            2973     138.0          2.115165        2.432543   
3              2474     144.0          2.524065        2.330842   
11            23343      59.0         -0.572456       -0.427050   

       infantMortality  lifeFemale  lifeMale  totalfertilityrate  
Index                                                             
1             1.247233   -1.198493 -0.895118            2.719042  
9            -0.333597    0.278282  0.387065            2.471951  
124           2.023276   -1.872458 -1.898566            2.348405  
3             3.374166   -2.427488 -2.277646            2.286632  
11            0.356220   -0.712842  0.030283            2.224859

#### 
Identify the African country, say country $A$, with the least Manhattan distance to country $G$.

*(8 points)*

In [71]:
# C2.5.2
# Getting the values of social indicators for Luxembourg
luxembourg_values = df[df['country'] == 'Luxembourg'][social_indicators].values[0]

df_africa = df[df['continent'] == 'Africa'].copy()

# Calculating the Manhattan distance between Luxembourg and each of the African countries
df_africa['manhattan_distance'] = df_africa[social_indicators].apply(
    lambda x: sum(abs(x - luxembourg_values)), axis=1)

df_africa[['country'] + social_indicators + ['manhattan_distance']].head()


country  economicActivityFemale  economicActivityMale  \
Index                                                               
124        Ethiopia                0.699967              1.053310   
146          Malawi                1.689691              0.521297   
68           Guinea               -0.090641              0.430466   
151    Burkina Faso                1.929802              1.598298   
155         Burundi                2.585714              1.754009   

       illiteracyFemale  illiteracyMale  infantMortality  lifeFemale  \
Index                                                                  
124            2.115165        2.432543         2.023276   -1.872458   
146            1.460132        0.853186         3.029258   -2.913138   
68             2.250142        2.169317         2.511896   -2.328375   
151            2.754319        3.389729         1.735852   -2.328375   
155            2.226322        2.205211         2.224472   -2.149973   

       lifeMale  totalfertilityrate  manhattan_distance  
Index                                                    
124   -1.898566            2.348405           21.506361  
146   -2.801669            2.156909           22.487953  
68    -2.166152            2.107491           20.935868  
151   -2.266497            2.082782           25.148325  
155   -2.221899            1.903640           24.333913

In [72]:
# C2.5.3
closest_country = df_africa.loc[df_africa['manhattan_distance'].idxmin()]['country']

closest_country


'Reunion'

The African country closest to Luxembourg, with regard to social indicators, is Reunion.

### 
Find the correlation between the Manhattan distance from country $G$ and GDP per capita for African countries.

*(6 points)*

In [73]:
# C2.6
correlation_af = df_africa['manhattan_distance'].corr(df_africa['gdpPerCapita'])
correlation_af


-0.7126967474772726

the correlation between the Manhattan distance from country $G$ and GDP per capita for African countries is -0.7126967474772726

### 
Based on the correlation coefficient in $2(f)$, do you think African countries should try to emulate the social characteristics of country $G$? Justify your answer.

*(4 points)*

## C2.7
The correlation between the Manhattan distance of social indicators from Luxembourg and GDP per capita for African countries is approximately −0.713. This negative correlation indicates that as the social indicators of an African country become more similar to those of Luxembourg (i.e., the Manhattan distance decreases), the GDP per capita of the African country tends to be higher. Luxembourg is a small, highly developed European country with a unique socio-economic context. African countries, on the other hand, have diverse socio-economic landscapes with different challenges and opportunities. Simply emulating Luxembourg’s social characteristics might not be practical or effective in the unique contexts of various African countries.

While the correlation suggests a relationship between similarity to Luxembourg’s social characteristics and higher GDP per capita, it might not be advisable for African countries to strictly emulate Luxembourg without considering their unique contexts and development priorities. A more nuanced, context-specific approach that considers various aspects of development might be more beneficial.

## Medical data

Read the data sets *conditions.csv and patients.csv*. Suppose we are interested in studying patients with prediabetes condition. Do not drop or compute any missing values. In *condition.csv*, the patient IDs are stored in column `PATIENT`, and the medical conditions are stored in column `DESCRIPTION`. In *patient.csv*, the patient IDs are stored in column `Id`.

In [78]:
# C3
conditions_df = pd.read_csv('/Users/jinwens/Downloads/conditions.csv')
patients_df = pd.read_csv('/Users/jinwens/Downloads/patients.csv')




(40     e9922ddd-8e85-9837-9ac8-885c34075450
 112    6d95ba48-a8be-b4c8-cdf6-f3d89142217b
 170    2599cde2-f690-8596-3cb5-905d701f6987
 182    527c52ef-be00-25b5-823b-b87005bf0d5f
 223    f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130
 Name: PATIENT, dtype: object,
 5)

### 
Print the patient IDs of all the patients with prediabetes condition.

*(4 points)*

In [81]:
# C3.1
prediabetes_patient_ids = conditions_df[conditions_df['DESCRIPTION'] == 'Prediabetes']['PATIENT']

prediabetes_patient_ids.head(), len(prediabetes_patient_ids)

(40     e9922ddd-8e85-9837-9ac8-885c34075450
 112    6d95ba48-a8be-b4c8-cdf6-f3d89142217b
 170    2599cde2-f690-8596-3cb5-905d701f6987
 182    527c52ef-be00-25b5-823b-b87005bf0d5f
 223    f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130
 Name: PATIENT, dtype: object,
 5)

Five patients have been identified with a prediabetes condition, and their IDs have been displayed.

### 
Make a subset of the data with only prediabetes patients. How many prediabetes patients are there?

*(4 points)*

**Hint:** `.isin()`

In [82]:
# C3.2
prediabetes_patients_df = patients_df[patients_df['Id'].isin(prediabetes_patient_ids)]

len(prediabetes_patients_df), prediabetes_patients_df.head()


(5,
                                      Id  Age  DEATHDATE          SSN  \
 1  6d95ba48-a8be-b4c8-cdf6-f3d89142217b   35        NaN  999-60-2854   
 3  527c52ef-be00-25b5-823b-b87005bf0d5f   54        NaN  999-40-6727   
 4  2599cde2-f690-8596-3cb5-905d701f6987   32        NaN  999-45-7333   
 6  e9922ddd-8e85-9837-9ac8-885c34075450   70        NaN  999-91-6093   
 9  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   55        NaN  999-92-5582   
 
      DRIVERS    PASSPORT PREFIX         FIRST        LAST  SUFFIX  ...  \
 1  S99933959  X72103595X    Mr.   Jefferey580  Murazik203     NaN  ...   
 3  S99951719   X1779287X   Mrs.    Valarie790   Hackett68     NaN  ...   
 4  S99931616  X84630753X    Mr.  Filiberto722     West559     NaN  ...   
 6  S99941771  X14585646X   Mrs.   Charlena802   Ernser583     NaN  ...   
 9  S99929701  X85339645X    Mr.      Bruno518   Lehner980     NaN  ...   
 
                       BIRTHPLACE              ADDRESS        CITY  \
 1      Boston  Massachusetts  US 

New subset of data is created above. 

### 
What proportion of the total `HEALTHCARE_EXPENSES` of all the patients correspond to the `HEALTHCARE_EXPENSES` of prediabetes patients.

*(4 points)*

In [83]:
# C3.3
total_expenses_all_patients = patients_df['HEALTHCARE_EXPENSES'].sum()
total_expenses_prediabetes = prediabetes_patients_df['HEALTHCARE_EXPENSES'].sum()

proportion_expenses_prediabetes = (total_expenses_prediabetes / total_expenses_all_patients) * 100

proportion_expenses_prediabetes

46.09622337553665

The HEALTHCARE_EXPENSES of prediabetes patients constitute approximately 46.10% of the total HEALTHCARE_EXPENSES of all patients in the dataset.

## Bonus question
This is an optional question with no partial credit. You will get points only if your solution is completely correct. We advise you to attempt it only when you are done with the rest of the assignment.

*(10 points)*

Read the file *STAT303-1 survey for data analysis.csv*. In this question, we'll work to clean this data a bit. As with every question, you are **not** allowed to use a `for` loop or any other loop.

Execute the following code to read the data and clean the column names.

In [98]:
survey_data = pd.read_csv('/Users/jinwens/Downloads/STAT303-1 survey for data analysis.csv')
new_col_names = ['parties_per_month', 'smoke', 'weed', 'introvert_extrovert', 'love_first_sight', 'learning_style', 'left_right_brained', 'personality_type', 'social_media', 'num_insta_followers', 'streaming_platforms', 'expected_marriage_age', 'expected_starting_salary', 'fav_sport', 'minutes_ex_per_week', 'sleep_hours_per_day', 'how_happy', 'farthest_distance_travelled', 'fav_number', 'fav_letter', 'internet_hours_per_day', 'only_child', 'birthdate_odd_even', 'birth_month', 'fav_season', 'living_location_on_campus', 'major', 'num_majors_minors', 'high_school_GPA', 'NU_GPA', 'age', 'height', 'height_father', 'height_mother', 'school_year', 'procrastinator', 'num_clubs', 'student_athlete', 'AP_stats', 'used_python_before', 'dominant_hand', 'childhood_in_US', 'gender', 'region_of_residence', 'political_affliation', 'cant_change_math_ability', 'can_change_math_ability', 'math_is_genetic', 'much_effort_is_lack_of_talent']
survey_data.columns = list(survey_data.columns[0:2])+new_col_names

Check the datatype of the variables using the `dtypes` attribute of the Pandas DataFrame object. You will notice that only two variables are numeric. However, if you check the first few observations of the data with the function `head()` you will find the there are several more variables that seem to have numeric values.

In [99]:
survey_data.dtypes, survey_data.head()

(Timestamp                         object
 fav_alcohol                       object
 parties_per_month                 object
 smoke                             object
 weed                              object
 introvert_extrovert               object
 love_first_sight                  object
 learning_style                    object
 left_right_brained                object
 personality_type                  object
 social_media                      object
 num_insta_followers               object
 streaming_platforms               object
 expected_marriage_age             object
 expected_starting_salary          object
 fav_sport                         object
 minutes_ex_per_week               object
 sleep_hours_per_day                int64
 how_happy                         object
 farthest_distance_travelled       object
 fav_number                        object
 fav_letter                        object
 internet_hours_per_day            object
 only_child                       

Write a function that accepts a Pandas Series *(or a column of a Pandas DataFrame object)* as argument, and if the datatype of the Series is non-numeric, does the following:

1. Checks if at least 10 values of the Series contain a digit in them.

2. If at least 10 values are found to contain a digit, then:

    A. Eliminate the characters `~`, `+`, and `,` from all the values of the Series.
    
    B. Convert the Series to numeric (with coercion if needed).
    
3. If at least 10 values are NOT found to contain a digit, then:

    A. If the values of the Series are *'Yes'* and *'No'*, then replace *'Yes'* with *1* and *'No'* with *0*. The Series datatype must change to numeric as well.
    
    B. If the values of the Series are *'Agree'* and *'Disagree'*, then replace *'Agree'* with *1* and *'Disagree'* with *0*. The Series datatype must change to numeric as well.
    
Apply the function to each column of `survey_data` using the *apply()* method. Save the updated DataFrame as `survey_data_clean`. Then, execute the following code.

In [101]:
def clean_column(column):
    # Convert all values in the column to strings
    str_column = column.astype(str)
    
    # Function to check if a value contains a digit
    def has_digit(value):
        return bool(re.search(r'\d', value))
    
    # Count how many values in the Series contain a digit
    digit_count = str_column.apply(has_digit).sum()
    
    # If at least 10 values contain a digit
    if digit_count >= 10:
        # Remove ~, +, and , characters, and convert to numeric, coercing errors
        return pd.to_numeric(str_column.str.replace('~|\+|,', '', regex=True), errors='coerce')
    
    # If values are 'Yes'/'No' or 'Agree'/'Disagree'
    else:
        return column.replace({'Yes': 1, 'No': 0, 'Agree': 1, 'Disagree': 0})

# Applying the function to each column except the first two (timestamp and participant_id)
survey_data_clean = survey_data.copy()
survey_data_clean.iloc[:, 2:] = survey_data_clean.iloc[:, 2:].apply(clean_column)

# Displaying the first few rows of the cleaned DataFrame
survey_data_clean.head()


/var/folders/vh/bnk8gjpn5998vq44hzdttvwc0000gn/T/ipykernel_14403/658712624.py:23: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  survey_data_clean.iloc[:, 2:] = survey_data_clean.iloc[:, 2:].apply(clean_column)


Timestamp              fav_alcohol  parties_per_month  \
0   2022/09/13 1:43:34 pm GMT-5            I don't drink                1.0   
1   2022/09/13 5:28:17 pm GMT-5  Hard liquor/Mixed drink                3.0   
2   2022/09/13 7:56:38 pm GMT-5  Hard liquor/Mixed drink                3.0   
3  2022/09/13 10:34:37 pm GMT-5  Hard liquor/Mixed drink               12.0   
4   2022/09/14 4:46:19 pm GMT-5            I don't drink                1.0   

  smoke          weed introvert_extrovert  love_first_sight  \
0     0  Occasionally           Introvert                 0   
1     0  Occasionally           Extrovert                 0   
2     0             1           Introvert                 0   
3     0             0           Extrovert                 0   
4     0             0           Extrovert                 1   

                                      learning_style  \
0  Visual (learn best through images or graphic o...   
1  Visual (learn best through images or graphic o...   
2  Kinesthetic (learn best through figuring out h...   
3  Visual (learn best through images or graphic o...   
4  Reading/Writing (learn best through words ofte...   

                                  left_right_brained personality_type  ...  \
0  Left-brained (logic, science, critical thinkin...             INFJ  ...   
1  Left-brained (logic, science, critical thinkin...             ESFJ  ...   
2  Left-brained (logic, science, critical thinkin...             ISTJ  ...   
3  Left-brained (logic, science, critical thinkin...             ENFJ  ...   
4  Right-brained (creative, art, imaginative, int...             ENTJ  ...   

  used_python_before  dominant_hand childhood_in_US  gender  \
0                  1          Right               1  Female   
1                  1          Right               1    Male   
2                  0          Right               0  Female   
3                  0          Right               1  Female   
4                  0          Right               1  Female   

   region_of_residence political_affliation  cant_change_math_ability  \
0            Northeast             Democrat                         0   
1                 West             Democrat                         0   
2        International       No affiliation                         0   
3            Southeast             Democrat                         0   
4            Northeast             Democrat                         1   

   can_change_math_ability math_is_genetic  much_effort_is_lack_of_talent  
0                        1               0                              0  
1                        1               0                              0  
2                        1               0                              0  
3                        1               0                              0  
4                        0               0                              0  

[5 rows x 51 columns]

In [102]:
survey_data_clean.describe().loc['mean',:]

parties_per_month                    5.052632
love_first_sight                     0.281250
num_insta_followers                969.211957
expected_marriage_age               29.185792
expected_starting_salary         85393.478723
minutes_ex_per_week                205.778378
sleep_hours_per_day                  7.223958
farthest_distance_travelled       6283.676271
fav_number                          37.978947
internet_hours_per_day              11.196809
only_child                           0.151042
num_majors_minors                    2.505263
high_school_GPA                      6.704000
NU_GPA                               3.755977
age                                 19.963351
height                              67.660495
height_father                       69.073518
height_mother                       63.377369
procrastinator                       0.640625
num_clubs                            2.593750
student_athlete                      0.015625
AP_stats                          

 The above code should print out the mean values of 28 numeric columns in the data `survey_data_clean`.

The variables that you should see as numeric in `survey_data_clean` are given in the list `numeric_columns` below *(this is just to check your work)*:

In [103]:
numeric_columns = ['parties_per_month', 'love_first_sight', 'num_insta_followers',
       'expected_marriage_age', 'expected_starting_salary',
       'minutes_ex_per_week', 'sleep_hours_per_day',
       'farthest_distance_travelled', 'fav_number', 'internet_hours_per_day',
       'only_child', 'num_majors_minors', 'high_school_GPA', 'NU_GPA', 'age',
       'height', 'height_father', 'height_mother', 'procrastinator',
       'num_clubs', 'student_athlete', 'AP_stats', 'used_python_before',
       'childhood_in_US', 'cant_change_math_ability',
       'can_change_math_ability', 'math_is_genetic',
       'much_effort_is_lack_of_talent']

Note that your **function must be general**, i.e., it must work for any other dataset as well. This means, you cannot hard code a column name, or anything specific to `survey_data` in the function.